### DISCUSSION TITLE
How to avoid re-building dependencies when only sources change?

### DISCUSSION THREAD

**Discussion by srid** [90-008/nix-cargo-integration#140] [Q&A] [answered]
## Problem

- Clone https://github.com/srid/rust-nix-template/pull/27/files (commit [1d00eeb](https://github.com/srid/rust-nix-template/pull/27/commits/1d00eeb11a325794bf5bcc8e3ff33ade1632318f))
- Run `nix build -L`
- Edit `main.rs` and save the new file
- Run `nix build -L` again

You'll see that the second nix build is recompiling all the dependency crates.

## Questions

- How can this be avoided? 
  - Doesn't nci use (via dream2nix) `crane` behind the scenes, with the appropriate `cargoArtifacts` separation[^1]?
- Is nci stable enough for use in real-world Rust projects? It appears that dream2nix is not; can the interface of nci be guaranteed to remain stable regardless what happens under the hood with nci?

---

[^1]: See https://crane.dev/examples/quick-start.html
    <img width="483" alt="image" src="https://github.com/yusdacra/nix-cargo-integration/assets/3998/f06f708c-0b2e-4e30-8164-adb1eb11f467">

**Comment by 90-008** [comment]
I don't see the dependency rebuilding behaviour on my end, seems to work as intended for me. I'm on Nix 2.17.0 if that helps.

> Doesn't nci use (via dream2nix) crane behind the scenes, with the appropriate cargoArtifacts separation

Yes, it does.

> Is nci stable enough for use in real-world Rust projects?

I've iterated over it a lot by now and I probably won't change the interface a lot after this, as dream2nix also won't change a lot since it switched to modules proper. I can't really promise considering it's not being used in bigger projects so it might not satisfy some needs perhaps and might need changes here and there but I believe most of these are probably going to be additions.

**Reply by srid** [reply]
> I don't see the dependency rebuilding behaviour on my end, seems to work as intended for me. I'm on Nix 2.17.0 if that helps.

Presumably because you are on Linux. I'm unable to reproduce this on NixOS either:

<img width="825" alt="image" src="https://github.com/yusdacra/nix-cargo-integration/assets/3998/a3a462f1-170e-40e3-b3be-8223a35cdd49">

But try it on macOS (ARM), where the dependencies will be rebuilt when you just change `src/main.rs`:

<img width="839" alt="image" src="https://github.com/yusdacra/nix-cargo-integration/assets/3998/a9648670-64d0-4fe5-8d80-1867e10f4275">

**Reply by srid** [reply]
Could this be related to using an older version of crane (nci uses 0.12)?

https://github.com/ipetkov/crane/blob/master/CHANGELOG.md

I tried upgrading but nci seems incompatible with the new version:

<img width="797" alt="image" src="https://github.com/yusdacra/nix-cargo-integration/assets/3998/6312e038-17f8-40f2-9406-e7edeaf50734">


```diff
diff --git a/flake.nix b/flake.nix
index 99382ef..c379344 100644
--- a/flake.nix
+++ b/flake.nix
@@ -9,6 +9,10 @@
     nci.inputs.nixpkgs.follows = "nixpkgs";
     nci.inputs.parts.follows = "flake-parts";
 
+    crane.url = "github:ipetkov/crane/v0.14.3";
+    crane.flake = false;
+    nci.inputs.crane.follows = "crane";
+
     # Dev tools
     treefmt-nix.url = "github:numtide/treefmt-nix";
   };

```

**Reply by srid** [reply]
Hmm, dream2nix itself is hardcoded to work with an older version of crane:

https://github.com/nix-community/dream2nix/blob/a900b3a332e36cf7a112283fdd1833b8311f9dda/modules/dream2nix/rust-crane/default.nix#L220-L225

Guess, I'm better off with using `crane` directly for my Rust projects ...

**Comment by 90-008** [comment, marked as answer]
> Hmm, dream2nix itself is hardcoded to work with an older version of crane:

It's not hardcoded, this is just a default. Everything specified in `deps` can be overrode, this is how NCI provides it's own crane source from it's flake inputs.

> Could this be related to using an older version of crane (nci uses 0.12)?

I think this issue is probably the same as https://github.com/nix-community/dream2nix/issues/506 then, i didn't have time to look into it though because i don't use non-linux systems. I will try updating to a newer crane version later and see if that fixes anything, though I think this might be an issue because of how crane is used in dream2nix itself rather than crane itself, but as i said i didn't look into that.
